src/state.rs

```diff
@@ -0,0 +1,185 @@
+        Ok(State {
+            network_name: SubnetID::from_str(&params.network_name)?,
+            atomic_exec_registry: TCid::new_hamt(store)?,
+            nonce: Default::default(),
+            ipc_gateway_address: Address::from_bytes(&params.ipc_gateway_address)?,
+            ipc_gateway_cross_method_num: MethodNum::from(params.ipc_gateway_cross_method_num),
+        })
+    }
+
+    /// Gets an atomic execution by cid from the state
+    pub fn get_atomic_exec<BS: Blockstore>(
+        &self,
+        store: &BS,
+        cid: &TCid<TLink<AtomicExecParams>>,
+    ) -> anyhow::Result<Option<AtomicExec>> {
+        let registry = self.atomic_exec_registry.load(store)?;
+        let exec = get_atomic_exec(&registry, cid)?;
+        Ok(exec.cloned())
+    }
+
+    /// Sets a new atomic exec with Cid
+    pub fn set_atomic_exec<BS: Blockstore>(
+        &mut self,
+        store: &BS,
+        cid: &TCid<TLink<AtomicExecParamsMeta>>,
+        exec: AtomicExec,
+    ) -> anyhow::Result<()> {
+        self.atomic_exec_registry.update(store, |registry| {
+            registry
+                .set(cid.cid().to_bytes().into(), exec)
+                .map_err(|e| e.downcast_wrap("failed to set atomic exec"))?;
+            Ok(())
+        })?;
+        Ok(())
+    }
+
+    pub fn rm_atomic_exec<BS: Blockstore>(
+        &mut self,
+        store: &BS,
+        cid: &TCid<TLink<AtomicExecParamsMeta>>,
+    ) -> anyhow::Result<()> {
+        self.atomic_exec_registry.update(store, |registry| {
+            registry
+                .delete(&cid.cid().to_bytes())
+                .map_err(|e| e.downcast_wrap("failed to delete atomic exec"))?;
+            Ok(())
+        })?;
+        Ok(())
+    }
+
+    /// Propagates the result of an execution to the corresponding subnets
+    /// in a cross-net message.
+    #[allow(clippy::too_many_arguments)]
+    pub fn propagate_exec_result<BS: Blockstore>(
+        &mut self,
+        store: &BS,
+        cid: &TCid<TLink<AtomicExecParamsMeta>>,
+        exec: &AtomicExec,
+        output: atomic::SerializedState, // LockableState to propagate. The same as in SubmitAtomicExecParams
+        _curr_epoch: ChainEpoch,
+        abort: bool,
+    ) -> anyhow::Result<Vec<(Address, MethodNum, RawBytes, TokenAmount)>> {
+        // This tracks the list of cross message to send
+        let mut msgs = vec![];
+
+        let mut visited = HashSet::new();
+        let params = exec.params();
+        for (k, v) in params.inputs.iter() {
+            let sn = k.subnet()?;
+            if visited.get(&sn).is_none() {
+                // let mut msg =
+                //     self.exec_result_msg(&sn, &v.actor, &params.msgs[0], output.clone(), abort)?;
+                let msg =
+                    self.exec_result_msg(&sn, &v.actor, &params.msgs[0], output.clone(), abort)?;
+
+                // TODO: is this cross message correct?
```

Comment:
Yes, this is the right way of doing it. That way we've decoupled the atomic exec completely from the `ipc_gateway`